admission-webhook/webhook.go

```diff
@@ -0,0 +1,402 @@
+		return nil, &podAdmissionError{error: fmt.Errorf("unable to unmarshall pod JSON object: %v", err), code: http.StatusBadRequest}
+	}
+
+	return pod, nil
+}
+
+// validateCreateRequest ensures that the only GMSA content annotations set on the pod,
+// match the corresponding GMSA name annotations, and that the pod's service account
+// is authorized to `use` the requested GMSA's.
+func (webhook *webhook) validateCreateRequest(pod *corev1.Pod, namespace string) (*admissionv1beta1.AdmissionResponse, *podAdmissionError) {
+	var err *podAdmissionError
+
+	iterateOverGMSAAnnotationPairs(pod, func(nameKey, contentsKey string) {
+		if err != nil {
+			return
+		}
+
+		if credSpecName, present := pod.Annotations[nameKey]; present && credSpecName != "" {
+			// let's check that the associated service account can read the relevant cred spec CRD
+			if authorized, reason := webhook.client.isAuthorizedToUseCredSpec(pod.Spec.ServiceAccountName, namespace, credSpecName); !authorized {
+				msg := fmt.Sprintf("service account %s is not authorized `use` gMSA cred spec %s", pod.Spec.ServiceAccountName, credSpecName)
+				if reason != "" {
+					msg += fmt.Sprintf(", reason: %q", reason)
+				}
+				err = &podAdmissionError{error: fmt.Errorf(msg), pod: pod, code: http.StatusForbidden}
+				return
+			}
+
+			// and the content annotation should contain the expected cred spec
+			if credSpecContents, present := pod.Annotations[contentsKey]; present {
+				if expectedContents, code, retrieveErr := webhook.client.retrieveCredSpecContents(credSpecName); retrieveErr != nil {
+					err = &podAdmissionError{error: retrieveErr, pod: pod, code: code}
+				} else if credSpecContents != expectedContents {
+					err = &podAdmissionError{error: fmt.Errorf("cred spec contained in annotation %s does not match the contents of GMSA %s", contentsKey, credSpecName), pod: pod, code: http.StatusForbidden}
+				}
+			}
+
+		} else if _, present := pod.Annotations[contentsKey]; present {
+			// the name annotation is not present, but the content one is
+			err = &podAdmissionError{error: fmt.Errorf("cannot pre-set a pod's gMSA content annotation (annotation %v present)", contentsKey), pod: pod, code: http.StatusForbidden}
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &admissionv1beta1.AdmissionResponse{Allowed: true}, nil
+}
+
+// mutateCreateRequest inlines the requested GMSA's into the pod's spec as annotations.
+func (webhook *webhook) mutateCreateRequest(pod *corev1.Pod) (*admissionv1beta1.AdmissionResponse, *podAdmissionError) {
+	var (
+		patches []map[string]string
+		err     *podAdmissionError
+	)
+
+	iterateOverGMSAAnnotationPairs(pod, func(nameKey, contentsKey string) {
+		if err != nil {
+			return
+		}
+
+		if _, present := pod.Annotations[contentsKey]; present {
+			// only this admission controller is allowed to populate the actual contents of the cred spec
+			// and "/mutate" is called before "/validate"
```

Comment:
Users should be allowed to specify the GMSA contents if they desire. While not typical usage, we should not error out if the contents happen to be specified by the user - we simply skip the population/mutation. Validation later on will make sure the populated values are in sync with CRD contents.

Reply:
Fair enough, didn't think of that.

If they do specify the contents though, they also need to specify the name; otherwise validation won't be able to assert whether the contents are legit.